kernel/cooking.ml

```diff
@@ -251,4 +256,115 @@ let cook_constant ~hcons { from = cb; info } =
+
+(* Replace
+
+     Var(y1)..Var(yq):C1..Cq |- Ij:Bj
+     Var(y1)..Var(yq):C1..Cq; I1..Ip:B1..Bp |- ci : Ti
+
+   by
+
+     |- Ij: (y1..yq:C1..Cq)Bj
+     I1..Ip:(B1 y1..yq)..(Bp y1..yq) |- ci : (y1..yq:C1..Cq)Ti[Ij:=(Ij y1..yq)]
+*)
+
+let it_mkNamedProd_wo_LetIn b d =
+  List.fold_left (fun c d -> mkNamedProd_wo_LetIn d c) b d
+
+let abstract_inductive decls nparamdecls inds =
+  let open Entries in
+  let ntyp = List.length inds in
+  let ndecls = Context.Named.length decls in
+  let args = Context.Named.to_instance mkVar (List.rev decls) in
+  let args = Array.of_list args in
+  let subs = List.init ntyp (fun k -> lift ndecls (mkApp(mkRel (k+1),args))) in
+  let inds' =
+    List.map
+      (function (tname,arity,template,cnames,lc) ->
+        let lc' = List.map (Vars.substl subs) lc in
+        let lc'' = List.map (fun b -> it_mkNamedProd_wo_LetIn b decls) lc' in
+        let arity' = it_mkNamedProd_wo_LetIn arity decls in
+        (tname,arity',template,cnames,lc''))
+        inds in
+  let nparamdecls' = nparamdecls + Array.length args in
+(* To be sure to be the same as before, should probably be moved to cook_inductive *)
+  let params' = let (_,arity,_,_,_) = List.hd inds' in
+                let (params,_) = decompose_prod_n_assum nparamdecls' arity in
+    params
+  in
+  let ind'' =
+  List.map
+    (fun (a,arity,template,c,lc) ->
+      let _, short_arity = decompose_prod_n_assum nparamdecls' arity in
+      let shortlc =
+        List.map (fun c -> snd (decompose_prod_n_assum nparamdecls' c)) lc in
+      { mind_entry_typename = a;
+        mind_entry_arity = short_arity;
+        mind_entry_template = template;
+        mind_entry_consnames = c;
+        mind_entry_lc = shortlc })
+    inds'
+  in (params',ind'')
+
+let refresh_polymorphic_type_of_inductive (_,mip) =
+  match mip.mind_arity with
+  | RegularArity s -> s.mind_user_arity, false
+  | TemplateArity ar ->
+    let ctx = List.rev mip.mind_arity_ctxt in
+      mkArity (List.rev ctx, Sorts.sort_of_univ ar.template_level), true
+
+let dummy_variance = let open Entries in function
+  | Monomorphic_entry _ -> assert false
+  | Polymorphic_entry (_,uctx) -> Array.make (Univ.UContext.size uctx) Univ.Variance.Irrelevant
+
+let cook_inductive { Opaqueproof.modlist; abstract } mib =
+  let open Entries in
+  let (section_decls, subst, abs_uctx) = abstract in
+  let nparamdecls = Context.Rel.length mib.mind_params_ctxt in
+  let subst, ind_univs =
+    match mib.mind_universes with
+    | Monomorphic ctx -> Univ.empty_level_subst, Monomorphic_entry ctx
+    | Polymorphic auctx ->
+      let subst, auctx = discharge_abstract_universe_context subst abs_uctx auctx in
+      let subst = Univ.make_instance_subst subst in
+      let nas = Univ.AUContext.names auctx in
+      let auctx = Univ.AUContext.repr auctx in
+      subst, Polymorphic_entry (nas, auctx)
+  in
+  let variance = match mib.mind_variance with
+    | None -> None
+    | Some _ -> Some (dummy_variance ind_univs)
```

Comment:
I don't know if dummy_variance is a great idea. It looks like it should produce a valid variance for the inductive, but if that was true it should use Invariant not Irrelevant. Maybe we should just leave the original variance and let rebuild deal with it (note that infercumul only tests whether it's a Some without looking at the content).

Reply:
I am not changing this in this PR, you should probably open a new PR atop of this simple merge.